wicket-extensions/src/main/java/org/apache/wicket/extensions/ajax/markup/html/modal/ModalDialog.java

```diff
@@ -0,0 +1,240 @@
+
+import com.github.openjson.JSONStringer;
+
+/**
+ * Presents a modal dialog to the user. See open and close methods.
+ * 
+ * @author Igor Vaynberg (ivaynberg)
+ */
+public class ModalDialog extends Panel
+{
+
+	public static final String CONTENT_ID = "content";
+
+	private final WebMarkupContainer container;
+	private final WebMarkupContainer contentContainer;
+	private boolean open = false;
+	private transient boolean openedInThisRequest = false;
+	private Options options;
+
+	public ModalDialog(String id)
+	{
+		super(id);
+		setOutputMarkupId(true);
+
+		// Container controls the overall visibility of the modal form innards. In its initial state
+		// this is set to
+		// hidden so only the external tag renders. When the modal is openedInThisRequest it is
+		// first repainted with
+		// this shown to
+		// insert the markup into the dom, and then the modal javascript is invoked to rip out this
+		// tag out of the dom
+		// and make it modal.
+		container = new WebMarkupContainer("container");
+		container.setOutputMarkupId(true);
+		container.setVisible(false);
+		add(container);
+
+		// We need this here in case the modal itself is placed inside a form. If that is the case
+		// and the content
+		// contains a form that form's markup form tag will be rendered as div because as far as
+		// wicket is concerned it
+		// is part of the same dom as the page and nested forms are forbidden. By overriding
+		// isRootForm() to true we are
+		// forcing this form - which will be ripped out of the dom along with the content form if
+		// there is one to always
+		// render its tag as 'form' instead of 'div'.
+		var form = new Form<Void>("form")
+		{
+			@Override
+			public boolean isRootForm()
+			{
+				return true;
+			}
+		};
+		form.setOutputMarkupId(true);
+		container.add(form);
+
+		contentContainer = form;
+		contentContainer.add(new EmptyPanel(CONTENT_ID));
+	}
+
+	@Override
+	public void renderHead(IHeaderResponse response)
+	{
+		super.renderHead(response);
+		response.render(JavaScriptHeaderItem.forReference(ModalDialogReferences.JS));
+
+		// if the page is refreshed and the window was open in the previous request we need to
+		// re-open it
+		if (open == true && openedInThisRequest == false)
+		{
+			response.render(OnDomReadyHeaderItem.forScript(getOpenJavascript()));
+		}
+	}
+
+	public ModalDialog open(AjaxRequestTarget target, WebMarkupContainer content)
+	{
+		open(target, null, content);
+		return this;
+	}
+
+	public ModalDialog open(AjaxRequestTarget target, Options options, WebMarkupContainer content)
+	{
+
+		if (!content.getId().equals(CONTENT_ID))
+		{
+			throw new IllegalArgumentException(
+				"Content must have wicket id set to ModalDialog.CONTENT_ID");
+		}
+
+		contentContainer.replace(content);
+
+		container.setVisible(true);
+
+		open = true;
+		openedInThisRequest = true;
+
+		target.add(this);
+
+		this.options = options;
+
+		target.prependJavaScript(getOpenJavascript());
```

Comment:
I would expect to use `append`JavaScript() here. I guess it works with `prepend` because of `setTimeout(..., 0)` below. Should we use `append` to be more clear ?

Reply:
the ```prepend``` is purposeful. if there are any other javascript contributed that works with elements that are inside the modal ```prepend``` guarantees that said javascript will execute after the affected elements are in the right place in the dom. i will add javadoc explaining this...

Reply:
I have to double check but I think the order is:
- prepend JS
- replace DOM
- append JS